src/source/Ice/NatBehaviorDiscovery.c

```diff
@@ -0,0 +1,404 @@
+/**
+ * Kinesis Video Producer Ice Utils
+ */
+#define LOG_CLASS "NatTypeDiscovery"
+#include "../Include_i.h"
+
+#define DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT                  5
+#define DEFAULT_TEST_NAT_TEST_RESPONSE_WAIT_TIME                    500 * HUNDREDS_OF_NANOS_IN_A_MILLISECOND
+
+#define NAT_BEHAVIOR_NONE_STR                                       (PCHAR) "NONE"
+#define NAT_BEHAVIOR_NOT_BEHIND_ANY_NAT_STR                         (PCHAR) "NOT_BEHIND_ANY_NAT"
+#define NAT_BEHAVIOR_NO_UDP_CONNECTIVITY_STR                        (PCHAR) "NO_UDP_CONNECTIVITY"
+#define NAT_BEHAVIOR_ENDPOINT_INDEPENDENT_STR                       (PCHAR) "ENDPOINT_INDEPENDENT"
+#define NAT_BEHAVIOR_ADDRESS_DEPENDENT_STR                          (PCHAR) "ADDRESS_DEPENDENT"
+#define NAT_BEHAVIOR_PORT_DEPENDENT_STR                             (PCHAR) "PORT_DEPENDENT"
+
+typedef struct {
+    /* Should be able to contain max number of binding response we can get */
+    PStunPacket bindingResponseList[DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT * 6];
+    UINT32 bindingResponseCount;
+    CVAR cvar;
+    MUTEX lock;
+} NatTestData, *PNatTestData;
+
+STATUS natTestIncomingDataHandler(
+        UINT64 customData, PSocketConnection pSocketConnection, PBYTE pBuffer, UINT32 bufferLen,
+        PKvsIpAddress pSrc, PKvsIpAddress pDest)
+{
+    UNUSED_PARAM(pSocketConnection);
+    UNUSED_PARAM(pSrc);
+    UNUSED_PARAM(pDest);
+
+    STATUS retStatus = STATUS_SUCCESS;
+    PNatTestData pNatTestData = (PNatTestData) customData;
+    PStunPacket pStunPacket = NULL;
+
+    MUTEX_LOCK(pNatTestData->lock);
+    if (pNatTestData->bindingResponseCount < DEFAULT_NAT_TEST_MAX_BINDING_REQUEST_COUNT * 3) {
+        CHK_STATUS(deserializeStunPacket(pBuffer, bufferLen, NULL, 0, &pStunPacket));
+        pNatTestData->bindingResponseList[pNatTestData->bindingResponseCount++] = pStunPacket;
+    }
+    MUTEX_UNLOCK(pNatTestData->lock);
+    CVAR_SIGNAL(pNatTestData->cvar);
+
+CleanUp:
+
+    return STATUS_SUCCESS;
+}
+
+STATUS executeNatTest(PStunPacket bindingRequest, PKvsIpAddress pDestAddr, PSocketConnection pSocketConnection,
+                      UINT32 testIndex, PNatTestData pData, PStunPacket *ppTestReponse)
+{
+    PStunPacket testResponse = NULL;
+    UINT32 i, j;
+    STATUS retStatus = STATUS_SUCCESS;
+
+    CHK(bindingRequest != NULL && pDestAddr != NULL && pSocketConnection != NULL && pData != NULL && ppTestReponse != NULL, STATUS_NULL_ARG);
+
+    MEMSET(bindingRequest->header.transactionId, 0x00, STUN_TRANSACTION_ID_LEN);
+    putInt32((PINT32) bindingRequest->header.transactionId, testIndex);
```

Comment:
Do you need to take care of the alignment here?

Reply:
no I dont. packet structure is defined in RFC.

Reply:
I am not asking for the rfc. I am asking to check whether the field is aligned (I think it is if I remember correctly) assuming the pointer to the structure is aligned and not a serialized buffer to an unaligned address

Reply:
the field is aligned: https://github.com/awslabs/amazon-kinesis-video-streams-webrtc-sdk-c/blob/master/src/source/Stun/Stun.h#L221